rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/WorkerBookKeeper.java

```diff
@@ -44,12 +154,58 @@ public void handleHeartbeat(String workerHostname, HeartbeatStatus request)
     throw new UnsupportedOperationException("Worker node should not handle heartbeat");
   }
 
+  @Override
+  public Map<String, NodeState> getClusterNodes()
+  {
+    RetryingBookkeeperClient client = threadLocalConnections.get();
+    try {
+      if (client == null) {
+        client = createBookKeeperClientWithRetry(bookKeeperFactory, masterHostname, conf);
```

Comment:
you are doing this already in getter of nodeStateMap cache, done need to check and set client here